pkg/client/query/cache/memory.go

```diff
@@ -0,0 +1,362 @@
+		return zero, ErrCacheMiss.Wrapf("key: %s, height: %d", key, getHeight)
+	}
+
+	return value.value, nil
+}
+
+// Set adds or updates the value in the cache for the given key. If the cache is
+// configured for historical mode, it will store the value at the latest **known**
+// height, which is only updated on calls to SetAtHeight, and therefore is not
+// guaranteed to be the current height.
+func (c *inMemoryCache[T]) Set(key string, value T) error {
+	if c.config.historical {
+		return c.SetAtHeight(key, value, c.latestHeight.Load())
+	}
+
+	isMaxKeysConfigured := c.config.maxKeys > 0
+	cacheHasMaxKeys := int64(len(c.values)) >= c.config.maxKeys
+	if isMaxKeysConfigured && cacheHasMaxKeys {
+		if err := c.evict(); err != nil {
+			return err
+		}
+	}
+
+	c.valuesMu.Lock()
+	defer c.valuesMu.Unlock()
+
+	c.values[key] = cacheValue[T]{
+		value:    value,
+		cachedAt: time.Now(),
+	}
+
+	return nil
+}
+
+// SetAtHeight adds or updates the historical value in the cache for the given key,
+// and at the given height. If the cache is not configured for historical mode, it
+// returns an error.
+func (c *inMemoryCache[T]) SetAtHeight(key string, value T, setHeight int64) error {
+	if !c.config.historical {
+		return ErrHistoricalModeNotEnabled
+	}
+
+	// Update c.latestHeight if the given setHeight is newer (higher).
+	latestHeight := c.latestHeight.Load()
+	if setHeight > latestHeight {
+		// NB: Only update if c.latestHeight hasn't changed since we loaded it above.
+		c.latestHeight.CompareAndSwap(latestHeight, setHeight)
+	}
+
+	c.valuesMu.Lock()
+	defer c.valuesMu.Unlock()
+
+	valueHistory, exists := c.valueHistories[key]
+	if !exists {
+		heightMap := make(map[int64]cacheValue[T])
+		valueHistory = cacheValueHistory[T]{
+			sortedDescHeights: make([]int64, 0),
+			heightMap:         heightMap,
+		}
+	}
+
+	// Update sortedDescHeights and ensure the list is sorted in descending order.
+	if _, setHeightExists := valueHistory.heightMap[setHeight]; !setHeightExists {
+		valueHistory.sortedDescHeights = append(valueHistory.sortedDescHeights, setHeight)
+		sort.Slice(valueHistory.sortedDescHeights, func(i, j int) bool {
+			return valueHistory.sortedDescHeights[i] > valueHistory.sortedDescHeights[j]
+		})
+	}
+
+	// Prune historical values for this key, where the setHeight
+	// is oder than the configured numHistoricalValues.
+	if c.config.numHistoricalValues > 0 {
+		lenCachedHeights := int64(len(valueHistory.sortedDescHeights))
+		for heightIdx := lenCachedHeights - 1; heightIdx >= 0; heightIdx-- {
+			cachedHeight := valueHistory.sortedDescHeights[heightIdx]
+
+			// DEV_NOTE: Since the list is sorted, and we're iterating from lowest
+			// (oldest) to highest (youngest) height, once we encounter a cachedHeight
+			// that is younger than the configured numHistoricalValues, ALL subsequent
+			// heights SHOULD also be younger than the configured numHistoricalValues.
+			if setHeight-cachedHeight <= c.config.numHistoricalValues {
```

Comment:
I struggle to make sense of this.

As of my understanding `config.numHistoricalValues` would be the "size" of the historical values retained. If so, the condition to delete an entry would be:
```
if
  len(valueHistory.sortedDescHeights) > config.numHistoricalValues &&
  setHeight.isYoungerThan(oldestCachedHeight)
```

Reply:
Great catch - thank you! :raised_hands: